src/shared/copy.c

```diff
@@ -1082,15 +1092,23 @@ static int fd_copy_tree_generic(
                 uid_t override_uid,
                 gid_t override_gid,
                 CopyFlags copy_flags,
+                const Set *denylist,
                 HardlinkContext *hardlink_context,
                 const char *display_path,
                 copy_progress_path_t progress_path,
                 copy_progress_bytes_t progress_bytes,
                 void *userdata) {
         int r;
 
+        if (set_contains(denylist, st)) {
+                log_debug("%s is in the denylist, skipping", from);
+                return 0;
+        }
```

Comment:
I am pretty sure this check should *not* be done for the top-level inode, i.e. the one we specify in CopyFiles=, but only for the children of it...

i.e. if I do `CopyFiles=/path/to/my/fstab:/etc/fstab /path/to/etc:/etc`

then this should copy in /etc/fstab from /path/to/my/fstab and everything else in /etc from /path/to/etc, but exclude fstab there.

Reply:
> I am pretty sure this check should not be done for the top-level inode, i.e. the one we specify in CopyFiles=, but only for the children of it...

I'm not sure why this shouldn't apply to the top-level inode from CopyFiles=. Generally, that's how people expect a denylist to work no? One caveat  with this approach is that the mountpoint for e.g. /home won't be available in the top level root partition, but that can easily be remedied with MakeDirectories.

> i.e. if I do CopyFiles=/path/to/my/fstab:/etc/fstab /path/to/etc:/etc

> then this should copy in /etc/fstab from /path/to/my/fstab and everything else in /etc from /path/to/etc, but exclude fstab there.

But this is for files in the same partition right? I don't see the point of having this mechanism apply to CopyFiles= from the same partition.  I would just rely on regular ordering there, in that files that are copied later override files that are copied earlier.

On top of that, we could still add CopyExcludeFiles= to add more directories to the denylist for a specific partition.

Reply:
well, so far things like COPY_SAME_MOUNT stop whenever we hit a mount point when descending, but not already if the top-level referenced inode is already one. I would always consider denylists/boundaries to apply when we go from one to the next, but never to the inode explicitly referenced.

I mean, if there was the need to also apply a denylist to the top-level inode we reference, then we could add a flag to enable such behaviour, but I doubt we'll ever need that.